### ISSUE TITLE
perf: critical path analysis for DistSQL

### ISSUE THREAD

**Issue by arjunravinarayan** [cockroachdb/cockroach#19288]
Perform a critical path analysis for DistSQL, similar to [the single key read critical path analysis](https://spencerkimball.github.io/critical-path/), focusing first on the simplest of DistSQL queries (`SELECT COUNT(*) FROM table`), to examine where the greatest performance bottlenecks are.

This will help triage performance work on making DistSQL faster.

**Comment by arjunravinarayan** [comment]
I'm working with a `COUNT(*)` query on the `lineitem` table from the
TPC-H scalefactor=10 dataset. Query run on a single node (a
`roachprod` worker with a local-ssd attached). The goal is to see
where the bottlenecks are in the system to running this query
faster. The query:

`SELECT COUNT(*) FROM lineitem`.

First step is to run the query (runtime: ~3m) and look at all the
profiles. The table's primary index is about 7GB in size, so that
gives us a rough end-to-end bandwidth of ~40MB/s.

Here is a [bare_cpuprofile.pdf](https://github.com/cockroachdb/cockroach/files/1409052/bare_cpuprofile.pdf).



The CPU profile shows that a healthy portion (~65%) of the work
happens under `MVCCIterate`, but 12% under `tablereader.go`, and 10%
under `aggregator.go`. It's hard to disentangle exactly what this
means from this graph, so I switched to a [bare_calltree.pdf](https://github.com/cockroachdb/cockroach/files/1409054/bare_calltree.pdf) profile. Here's a
better picture: `aggregator.go` takes 10% of CPU, of which 4% is
memory allocation, and 4% is go channel traffic. Second,
`tablereader.go` is 14% of CPU, of which 4% is go channel traffic,
and 3% is in finalizing `EncDatumRow`s in `RowFetcher`.

Looking at heap profiles (`inuse_space`, `inuse_objects`,
`alloc_space`, `alloc_objects`) doesn't reveal anything significant at
the distsql transport layer. So it's not quite that DistSQL is
allocating huge amounts of memory, but that it's getting stuck
transporting it around.

I made `kvFetcher` throw rows away and performance improved by 20%,
which shows that improvements on the DistSQL side can only give us at
most that much throughput increase. Currently working on getting some
more visibility at the kvFetcher layer and below.

To investigate what I said earlier about "go channel traffic" : to
execute a `COUNT(*)`, tablereader transports the rows (one by one) via
a `RowChannel` to its destination processor (in this case,
`aggregator`). The router of choice on local machines is a
`RowChannel`. I wrote a benchmark (see #19471) to see how fast
RowChannels can go, and they appear bottlenecked at ~50MB/s. Since the
kvFetcher is not really cranking out more performance than that, this
isn't a real bottleneck, but it could be down the line. 

The next step is to investigate if we can get more bandwidth out of `RowFetcher`, and how that compares to the bandwidth we get out of RocksDB running on a bare disk.

**Comment by arjunravinarayan** [comment]
This is one archetype (distsql performance in sending rows in ways that are typical of transfers within a single machine). I'm also working on a critical path analysis of common operations that always go across multiple machines (I'm working with a simple `JOIN` query across two tables on a primary key column) that would be more bottlenecked on GRPC `Send`. That also has the interesting property of requiring `EncDatum` decodings, since the rows do need decoding.

**Comment by petermattis** [comment]
Go CPU profiles sample running goroutines. If a goroutine is blocked on a mutex or channel or waiting for the network it won't show up in a CPU profile. I point this out so that you can be wary of drawing too many conclusions from the CPU profiles. Also take a look at `go tool trace` and blocking profiles (which I've so far not found terribly useful).

**Comment by arjunravinarayan** [comment]
The mutex profiles and block profiles didn't show anything interesting. I checked.

**Comment by asubiotto** [comment]
NOTE: All numbers from a one node cluster running on my laptop with key/value size ~200bytes (similar to tpch.lineitem).

To shed some more light on the layers below the `TableReader` (see https://github.com/cockroachdb/cockroach/pull/19602 for experiments regarding `RocksDB` iteration), I experimented with having the MVCCScan layer use an iterator that returns the same in-memory key and value over and over again to get some more insight into how these layers behaved with no storage bottleneck.

I specifically looked at `kvfetcher` and `MVCCScan`. First though, I benchmarked how quickly the experimental iterator ran and got ~`21000MB/s`.

Then, the `MVCCScan` layer was tested in batches of 10000 keys (similar to how the `kvfetcher` issues `fetch` calls), resulting in ~`520MB/s`. The `kvfetcher` ran at about `473MB/s`.

These numbers show that the layers result in a pretty big throughput hit to the straightforward iteration. However, the difference of the two layers is not big in relative terms (~10% throughput hit vs orders of magnitude) which suggests that the biggest bottleneck is the MVCCScan layer. We should rerun these experiments once we have the RocksDB layer returning `WriteBatches` and take a look how the throughput of these layers changes as a result of this.
